src/robot/running/arguments/typeconverters.py

```diff
@@ -389,61 +389,165 @@ class ListConverter(TypeConverter):
     type_name = 'list'
     abc = Sequence
     value_types = (str, Sequence)
+    aliases = ('list',)
 
-    def no_conversion_needed(self, value):
-        if isinstance(value, str):
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
+
+    @classmethod
+    def handles(cls, type_):
+        base_type = getattr(type_, '__origin__', type_)
+        handled = (cls.type, cls.abc) if cls.abc else cls.type
+        if base_type is tuple:
             return False
-        return super().no_conversion_needed(value)
+        return isinstance(base_type, type) and issubclass(base_type, handled)
+
+    def no_conversion_needed(self, value):
+        return not self.nested_types \
+               and isinstance(value, self.used_type) \
+               and not isinstance(value, str)
 
     def _non_string_convert(self, value, explicit_type=True):
-        return list(value)
+        if isinstance(value, str):
+            converted_list = self._literal_eval(value, list)
+        elif is_list_like(value):
+            converted_list = list(value)
+        else:
+            raise ValueError
+        if self.nested_types:
+            if TypeConverter.converter_for(self.nested_types[0],
+                                           languages=self.languages):
+                for i, elem in enumerate(converted_list) :
+                    converter = TypeConverter.converter_for(self.nested_types[0],
+                                                            languages=self.languages)
```

Comment:
With the max line length limitation I have started using way shorter names than I generally like. Not an issue here though. I changed to index.

Reply:
I took the converter out of the loop. The question about List[int, str] is a good one. I don't have a straight answer. My first thought it that you should use a tuple instead of a list when mixing types. Not sure what Python supports.